src/gui/MidiClipView.cpp

```diff
@@ -140,8 +142,61 @@ void MidiClipView::changeName()
 
 
 
+void MidiClipView::transposeSelection()
+{
+	const auto selection = getClickedClips();
+
+	// Calculate the key boundries for all clips
+	int highest = 0;
+	int lowest = NumKeys - 1;
+	for (ClipView* clipview: selection)
+	{
+		if (auto mcv = dynamic_cast<MidiClipView*>(clipview))
+		{
+			TimePos start, end;
+			int low, high;
+			if (mcv->getMidiClip()->notes().getBounds(start, end, low, high))
+			{
+				lowest = std::min(low, lowest);
+				highest = std::max(high, highest);
+			}
+		}
+	}
+
+	int minStep = 0 - lowest;
+	int maxStep = NumKeys - 1 - highest;
+	int semitones = QInputDialog::getInt(this, tr("Transpose"), tr("Semitones to transpose:"), 0, minStep, maxStep, 1);
+	if (semitones == 0) { return; }
+
+	// TODO make this not crash
+	// Engine::getSong()->addJournalCheckPoint();
+
+	QSet<Track*> m_changedTracks;
+	for (ClipView* clipview: getClickedClips())
+	{
+		if (auto mcv = dynamic_cast<MidiClipView*>(clipview))
+		{
+			auto track = clipview->getTrackView()->getTrack();
+			if (!m_changedTracks.contains(track))
+			{
+				track->addJournalCheckPoint();
+				m_changedTracks.insert(track);
+			}
+			auto clip = mcv->getMidiClip();
+			clip->notes().transpose(semitones);
+			emit clip->dataChanged();
```

Comment:
Is that line equal to
```c++
update();
if( getGUI()->pianoRoll()->currentMidiClip() == m_clip )
{
	getGUI()->pianoRoll()->update();
}
```
? I really can't find out where that signal leads to. At least, the piano roll is being refreshed now.

Reply:
Yes, it updates both the clip view and piano roll if active